Promoting GADT field selectors fails with "Unable to match type" error

I would expect this to work:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE TemplateHaskell #-}
module Bug where

import Data.Kind
import Data.Singletons.TH

$(promote [d|
  data Nat = Z | S Nat

  data Vec :: Nat -> Type -> Type where
    Nil :: Vec Z a
    (:>) :: { head :: a, tail :: Vec n a } -> Vec (S n) a
  |])
```

But if I try compiling this, I get the following inscrutable error message:

```
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )

Bug.hs:9:3: error:
    Unable to match type (DAppT (DAppT (DConT Vec_6989586621679040818) (DAppT (DConT S_6989586621679040817) (DVarT n_6989586621679040825))) (DVarT a_6989586621679040824)) with (DAppT (DAppT (DConT Vec_6989586621679040818) (DVarT a_6989586621679040826)) (DVarT a_6989586621679040827))
  |
9 | $(promote [d|
  |   ^^^^^^^^^^^...

Bug.hs:9:3: error: Q monad failure
  |
9 | $(promote [d|
  |   ^^^^^^^^^^^...
```

Some quick digging through the code reveals that [`conExistentialTvbs`](https://github.com/goldfirere/th-desugar/blob/612e58eacea0ef71f9f2ffbb1ae9305e77e303e3/Language/Haskell/TH/Desugar.hs#L412-L428) in `th-desugar` is the culprit, as it attempts to invoke `matchTy (Vec (S n) a) (Vec n a)`. If anything, I would have expected it to be the other way around, so I wonder if simply swapping the order of arguments to that call to `matchTy` would fix the issue.

--- a/.travis.yml
+++ b/.travis.yml
@@ -90,7 +90,7 @@ install:
     echo "source-repository-package"                            >> cabal.project
     echo "  type:     git"                                      >> cabal.project
     echo "  location: https://github.com/goldfirere/th-desugar" >> cabal.project
-    echo "  tag:      f075206882ce4e554c37537e624b4be7409d74a3" >> cabal.project
+    echo "  tag:      5eef50aad9a7db9ad284a427b791891398304fa8" >> cabal.project
     echo ""                                                     >> cabal.project
     echo "package th-desugar"                                   >> cabal.project
     echo "  tests:      False"                                  >> cabal.project
@@ -124,7 +124,7 @@ script:
     echo "source-repository-package"                            >> cabal.project
     echo "  type:     git"                                      >> cabal.project
     echo "  location: https://github.com/goldfirere/th-desugar" >> cabal.project
-    echo "  tag:      f075206882ce4e554c37537e624b4be7409d74a3" >> cabal.project
+    echo "  tag:      5eef50aad9a7db9ad284a427b791891398304fa8" >> cabal.project
     echo ""                                                     >> cabal.project
     echo "package th-desugar"                                   >> cabal.project
     echo "  tests:      False"                                  >> cabal.project

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -46,6 +46,28 @@ Changelog for singletons project
       natMinus a@(S _) Z     = a
       |])
     ```
+* The specification for how `singletons` deals with record selectors has been
+  simplified. Previously, `singletons` would try to avoid promoting so-called
+  "naughty" selectors (those whose types mention existential type variables
+  that do not appear in the constructor's return type) to top-level functions.
+  Determing if a selector is naughty is quite challenging in practice, as
+  determining if a type variable is existential or not in the context of
+  Template Haskell is difficult in the general case. As a result, `singletons`
+  now adopts the dumb-but-predictable approach of always promoting record
+  selectors to top-level functions, naughty or not.
+
+  This means that attempting to promote code with a naughty record selector,
+  like in the example below, will no longer work:
+
+  ```hs
+  $(promote [d|
+    data Some :: (Type -> Type) -> Type where
+      MkSome :: { getSome :: f a } -> Some f
+      -- getSome is naughty due to mentioning the type variable `a`
+    |])
+  ```
+
+  Please open an issue if you find this restriction burdensome in practice.
 
 2.7
 ---

--- a/README.md
+++ b/README.md
@@ -794,6 +794,26 @@ since `singletons` desugars away most uses of record syntax. On the other hand,
 it is not possible to write out code like
 `SIdentity { sRunIdentity = SIdentity STrue }` by hand.
 
+Another caveat is that GHC allows defining so-called "naughty" record selectors
+that mention existential type variables that do not appear in the constructor's
+return type. Naughty record selectors can be used in pattern matching, but they
+cannot be used as top-level functions. Here is one example of a naughty
+record selector:
+
+```hs
+data Some :: (Type -> Type) -> Type where
+  MkSome :: { getSome :: f a } -> Some f
+```
+
+Because `singletons` promotes all records to top-level functions, however,
+attempting to promote `getSome` will result in an invalid definition. (It
+may typecheck, but it will not behave like you would expect.) Theoretically,
+`singletons` could refrain from promoting naughty record selectors, but this
+would require detecting which type variables in a data constructor are
+existentially quantified. This is very challenging in general, so we stick to
+the dumb-but-predictable approach of always promoting record selectors,
+regardless of whether they are naughty or not.
+
 ### Signatures in patterns
 
 `singletons` can promote basic pattern signatures, such as in the following

--- a/cabal.project
+++ b/cabal.project
@@ -3,4 +3,4 @@ packages: .
 source-repository-package
   type: git
   location: https://github.com/goldfirere/th-desugar
-  tag: f075206882ce4e554c37537e624b4be7409d74a3
+  tag: 5eef50aad9a7db9ad284a427b791891398304fa8

--- a/singletons.cabal
+++ b/singletons.cabal
@@ -58,7 +58,7 @@ library
                       ghc-boot-th,
                       template-haskell,
                       containers >= 0.5,
-                      th-desugar >= 1.11 && < 1.12,
+                      th-desugar >= 1.12 && < 1.13,
                       pretty,
                       syb >= 0.4,
                       text >= 1.2,

--- a/src/Data/Singletons/Deriving/Enum.hs
+++ b/src/Data/Singletons/Deriving/Enum.hs
@@ -25,13 +25,21 @@ import Data.Maybe
 
 -- monadic for failure only
 mkEnumInstance :: DsMonad q => DerivDesc q
-mkEnumInstance mb_ctxt ty (DataDecl data_name tvbs cons) = do
-  let data_ty = foldTypeTvbs (DConT data_name) tvbs
-  non_vanilla <- isNonVanillaDataType data_ty cons
+mkEnumInstance mb_ctxt ty (DataDecl _ _ cons) = do
+  -- GHC only allows deriving Enum instances for enumeration types (i.e., those
+  -- data types whose constructors all lack fields). We perform the same
+  -- validity check here.
+  --
+  -- GHC actually goes further than we do. GHC will give a specific error
+  -- message if you attempt to derive an instance for a "non-vanilla" data
+  -- type—that is, a data type that uses features not expressible with
+  -- Haskell98 syntax, such as existential quantification. Checking whether
+  -- a type variable is existentially quantified is difficult in Template
+  -- Haskell, so we omit this check.
   when (null cons ||
-        any (\(DCon _ _ _ f _) ->
-              non_vanilla || not (null $ tysOfConFields f)) cons) $
+        any (\(DCon _ _ _ f _) -> not (null $ tysOfConFields f)) cons) $
     fail ("Can't derive Enum instance for " ++ pprint (typeToTH ty) ++ ".")
+
   n <- qNewName "n"
   let to_enum = UFunction [DClause [DVarP n] (to_enum_rhs cons [0..])]
       to_enum_rhs [] _ = DVarE errorName `DAppE` DLitE (StringL "toEnum: bad argument")

--- a/src/Data/Singletons/Deriving/Util.hs
+++ b/src/Data/Singletons/Deriving/Util.hs
@@ -16,7 +16,6 @@
 module Data.Singletons.Deriving.Util where
 
 import Control.Monad
-import qualified Data.List as List
 import Data.Singletons.Names
 import Data.Singletons.Syntax
 import Data.Singletons.Util
@@ -31,56 +30,6 @@ type DerivDesc q
   -> DType       -- The data type argument to the class.
   -> DataDecl    -- The original data type information.
   -> q UInstDecl -- The derived instance.
-
--- | Is this data type a non-vanilla data type? Here, \"non-vanilla\" refers to
--- any data type that cannot be expressed using Haskell98 syntax. For instance,
--- this GADT:
---
--- @
--- data Foo :: Type -> Type where
---   MkFoo :: forall a. a -> Foo a
--- @
---
--- Is equivalent to this Haskell98 data type:
---
--- @
--- data Foo a = MkFoo a
--- @
---
--- However, the following GADT is non-vanilla:
---
--- @
--- data Bar :: Type -> Type where
---   MkBar :: Int -> Bar Int
--- @
---
--- Since there is no equivalent Haskell98 data type. The closest you could get
--- is this:
---
--- @
--- data Bar a = (a ~ Int) => MkBar Int
--- @
---
--- Which requires language extensions to write.
---
--- A data type is a non-vanilla if one of the following conditions are met:
---
--- 1. A constructor has any existentially quantified type variables.
---
--- 2. A constructor has a context.
---
--- We care about this because some derivable stock classes, such as 'Enum',
--- forbid derived instances for non-vanilla data types.
-isNonVanillaDataType :: forall q. DsMonad q => DType -> [DCon] -> q Bool
-isNonVanillaDataType data_ty = anyM $ \con@(DCon _ ctxt _ _ _) -> do
-    ex_tvbs <- conExistentialTvbs data_ty con
-    return $ not $ null ex_tvbs && null ctxt
-  where
-    anyM :: (a -> q Bool) -> [a] -> q Bool
-    anyM _ [] = return False
-    anyM p (x:xs) = do
-      b <- p x
-      if b then return True else anyM p xs
 
 -----
 -- Utilities for deriving Functor-like classes.
@@ -204,8 +153,24 @@ isInTypeFamilyApp name tyFun tyArgs =
 -- A crude approximation of cond_functorOK from GHC. This checks that:
 --
 -- (1) There's at least one type variable in the data type.
--- (2) It doesn't use the last type variable in the wrong place, e.g. data T a = MkT (X a a)
--- (3) It doesn't constrain the last type variable, e.g., data T a = Eq a => MkT a
+-- (2) It doesn't constrain the last type variable, e.g., data T a = Eq a => MkT a
+-- (3) It doesn't use the last type variable in the wrong place, e.g. data T a = MkT (X a a)
+--
+-- This skips some things that cond_functorOK checks for but are tricky to
+-- implement in Template Haskell, such as if the last type variable in the
+-- constructor's return type is universally quantified. For example,
+-- functorLikeValidityChecks would accept the following example that
+-- cond_functorOK would reject:
+--
+-- @
+-- data T a b where
+--   MkT :: z -> T z z -- Last type variable is existential
+-- deriving instance Functor (T a)
+-- @
+--
+-- This isn't the end of the world, as it just means that the user will have to
+-- deal with a more complex error message when the generate code fails to
+-- typecheck.
 functorLikeValidityChecks :: forall q. DsMonad q => Bool -> DataDecl -> q ()
 functorLikeValidityChecks allowConstrainedLastTyVar (DataDecl n data_tvbs cons)
   | null data_tvbs -- (1)
@@ -221,16 +186,13 @@ functorLikeValidityChecks allowConstrainedLastTyVar (DataDecl n data_tvbs cons)
 
     -- (2)
     check_universal :: DCon -> q ()
-    check_universal con@(DCon con_tvbs con_theta con_name _ res_ty)
+    check_universal (DCon _ con_theta con_name _ res_ty)
       | allowConstrainedLastTyVar
       = pure ()
       | (_, res_ty_args) <- unfoldDType res_ty
       , (_, last_res_ty_arg) <- snocView $ filterDTANormals res_ty_args
       , Just last_tv <- getDVarTName_maybe last_res_ty_arg
-      = do ex_tvbs <- conExistentialTvbs (foldTypeTvbs (DConT n) data_tvbs) con
-           let univ_tvb_names = map extractTvbName con_tvbs List.\\ map extractTvbName ex_tvbs
-           if last_tv `elem` univ_tvb_names
-                && last_tv `OSet.notMember` foldMap fvDType con_theta
+      = do if last_tv `OSet.notMember` foldMap fvDType con_theta
               then pure ()
               else fail $ badCon con_name existential
       | otherwise

--- a/src/Data/Singletons/Promote.hs
+++ b/src/Data/Singletons/Promote.hs
@@ -287,12 +287,11 @@ promoteDataDecs = concatMapM promoteDataDec
 --    be promoted in a single location.
 --    See Note [singletons and record selectors] in D.S.Single.Data.
 promoteDataDec :: DataDecl -> PrM [DLetDec]
-promoteDataDec (DataDecl data_name tvbs ctors) = do
-  let arg_ty        = foldTypeTvbs (DConT data_name) tvbs
-      rec_sel_names = nub $ concatMap extractRecSelNames ctors
+promoteDataDec (DataDecl _ _ ctors) = do
+  let rec_sel_names = nub $ concatMap extractRecSelNames ctors
                       -- Note the use of nub: the same record selector name can
                       -- be used in multiple constructors!
-  rec_sel_let_decs <- getRecordSelectors arg_ty ctors
+  rec_sel_let_decs <- getRecordSelectors ctors
   ctorSyms         <- buildDefunSymsDataD ctors
   infix_decs       <- promoteReifiedInfixDecls rec_sel_names
   emitDecs $ ctorSyms ++ infix_decs

--- a/tests/SingletonsTestSuite.hs
+++ b/tests/SingletonsTestSuite.hs
@@ -131,6 +131,7 @@ tests =
     , compileAndDumpStdTest "T410"
     , compileAndDumpStdTest "T412"
     , compileAndDumpStdTest "T414"
+    , compileAndDumpStdTest "T443"
     , afterSingletonsNat .
       compileAndDumpStdTest "T445"
     , compileAndDumpStdTest "T453"

--- a/tests/compile-and-dump/Singletons/T443.golden
+++ b/tests/compile-and-dump/Singletons/T443.golden
@@ -1,0 +1,118 @@
+Singletons/T443.hs:(0,0)-(0,0): Splicing declarations
+    withOptions defaultOptions {genSingKindInsts = False}
+      $ singletons
+          $ lift
+              [d| data Nat = Z | S Nat
+                  data Vec :: Nat -> Type -> Type
+                    where
+                      VNil :: Vec Z a
+                      (:>) :: {head :: a, tail :: Vec n a} -> Vec (S n) a |]
+  ======>
+    data Nat = Z | S Nat
+    data Vec :: Nat -> Type -> Type
+      where
+        VNil :: Vec 'Z a
+        (:>) :: {head :: a, tail :: (Vec n a)} -> Vec ('S n) a
+    type ZSym0 :: Nat
+    type family ZSym0 where
+      ZSym0 = Z
+    type SSym0 :: (~>) Nat Nat
+    data SSym0 a0123456789876543210
+      where
+        SSym0KindInference :: SameKind (Apply SSym0 arg) (SSym1 arg) =>
+                              SSym0 a0123456789876543210
+    type instance Apply SSym0 a0123456789876543210 = S a0123456789876543210
+    instance SuppressUnusedWarnings SSym0 where
+      suppressUnusedWarnings = snd (((,) SSym0KindInference) ())
+    type SSym1 :: Nat -> Nat
+    type family SSym1 a0123456789876543210 where
+      SSym1 a0123456789876543210 = S a0123456789876543210
+    type VNilSym0 :: Vec Z a
+    type family VNilSym0 where
+      VNilSym0 = VNil
+    type (:>@#@$) :: (~>) a ((~>) (Vec n a) (Vec (S n) a))
+    data (:>@#@$) a0123456789876543210
+      where
+        (::>@#@$###) :: SameKind (Apply (:>@#@$) arg) ((:>@#@$$) arg) =>
+                        (:>@#@$) a0123456789876543210
+    type instance Apply (:>@#@$) a0123456789876543210 = (:>@#@$$) a0123456789876543210
+    instance SuppressUnusedWarnings (:>@#@$) where
+      suppressUnusedWarnings = snd (((,) (::>@#@$###)) ())
+    type (:>@#@$$) :: a -> (~>) (Vec n a) (Vec (S n) a)
+    data (:>@#@$$) a0123456789876543210 a0123456789876543210
+      where
+        (::>@#@$$###) :: SameKind (Apply ((:>@#@$$) a0123456789876543210) arg) ((:>@#@$$$) a0123456789876543210 arg) =>
+                         (:>@#@$$) a0123456789876543210 a0123456789876543210
+    type instance Apply ((:>@#@$$) a0123456789876543210) a0123456789876543210 = (:>) a0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings ((:>@#@$$) a0123456789876543210) where
+      suppressUnusedWarnings = snd (((,) (::>@#@$$###)) ())
+    type (:>@#@$$$) :: a -> Vec n a -> Vec (S n) a
+    type family (:>@#@$$$) a0123456789876543210 a0123456789876543210 where
+      (:>@#@$$$) a0123456789876543210 a0123456789876543210 = (:>) a0123456789876543210 a0123456789876543210
+    type TailSym0 :: (~>) (Vec (S n) a) (Vec n a)
+    data TailSym0 a0123456789876543210
+      where
+        TailSym0KindInference :: SameKind (Apply TailSym0 arg) (TailSym1 arg) =>
+                                 TailSym0 a0123456789876543210
+    type instance Apply TailSym0 a0123456789876543210 = Tail a0123456789876543210
+    instance SuppressUnusedWarnings TailSym0 where
+      suppressUnusedWarnings = snd (((,) TailSym0KindInference) ())
+    type TailSym1 :: Vec (S n) a -> Vec n a
+    type family TailSym1 a0123456789876543210 where
+      TailSym1 a0123456789876543210 = Tail a0123456789876543210
+    type HeadSym0 :: (~>) (Vec (S n) a) a
+    data HeadSym0 a0123456789876543210
+      where
+        HeadSym0KindInference :: SameKind (Apply HeadSym0 arg) (HeadSym1 arg) =>
+                                 HeadSym0 a0123456789876543210
+    type instance Apply HeadSym0 a0123456789876543210 = Head a0123456789876543210
+    instance SuppressUnusedWarnings HeadSym0 where
+      suppressUnusedWarnings = snd (((,) HeadSym0KindInference) ())
+    type HeadSym1 :: Vec (S n) a -> a
+    type family HeadSym1 a0123456789876543210 where
+      HeadSym1 a0123456789876543210 = Head a0123456789876543210
+    type Tail :: Vec (S n) a -> Vec n a
+    type family Tail a where
+      Tail ((:>) _ field) = field
+    type Head :: Vec (S n) a -> a
+    type family Head a where
+      Head ((:>) field _) = field
+    sTail ::
+      forall n a (t :: Vec (S n) a).
+      Sing t -> Sing (Apply TailSym0 t :: Vec n a)
+    sHead ::
+      forall n a (t :: Vec (S n) a).
+      Sing t -> Sing (Apply HeadSym0 t :: a)
+    sTail ((:%>) _ (sField :: Sing field)) = sField
+    sHead ((:%>) (sField :: Sing field) _) = sField
+    instance SingI (TailSym0 :: (~>) (Vec (S n) a) (Vec n a)) where
+      sing = (singFun1 @TailSym0) sTail
+    instance SingI (HeadSym0 :: (~>) (Vec (S n) a) a) where
+      sing = (singFun1 @HeadSym0) sHead
+    data SNat :: Nat -> Type
+      where
+        SZ :: SNat (Z :: Nat)
+        SS :: forall (n :: Nat). (Sing n) -> SNat (S n :: Nat)
+    type instance Sing @Nat = SNat
+    data SVec :: forall (a :: Nat) (a :: Type). Vec a a -> Type
+      where
+        SVNil :: forall a. SVec (VNil :: Vec Z a)
+        (:%>) :: forall a n (n :: a) (n :: Vec n a).
+                 (Sing n) -> (Sing n) -> SVec ((:>) n n :: Vec (S n) a)
+    type instance Sing @(Vec a a) = SVec
+    instance SingI Z where
+      sing = SZ
+    instance SingI n => SingI (S (n :: Nat)) where
+      sing = SS sing
+    instance SingI (SSym0 :: (~>) Nat Nat) where
+      sing = (singFun1 @SSym0) SS
+    instance SingI VNil where
+      sing = SVNil
+    instance (SingI n, SingI n) =>
+             SingI ((:>) (n :: a) (n :: Vec n a)) where
+      sing = ((:%>) sing) sing
+    instance SingI ((:>@#@$) :: (~>) a ((~>) (Vec n a) (Vec (S n) a))) where
+      sing = (singFun2 @(:>@#@$)) (:%>)
+    instance SingI d =>
+             SingI ((:>@#@$$) (d :: a) :: (~>) (Vec n a) (Vec (S n) a)) where
+      sing = (singFun1 @((:>@#@$$) (d :: a))) ((:%>) (sing @d))

--- a/tests/compile-and-dump/Singletons/T443.hs
+++ b/tests/compile-and-dump/Singletons/T443.hs
@@ -1,0 +1,15 @@
+module T443 where
+
+import Control.Monad.Trans.Class
+import Data.Kind
+import Data.Singletons.TH
+import Data.Singletons.TH.Options
+
+$(withOptions defaultOptions{genSingKindInsts = False} $
+  singletons $ lift [d|
+  data Nat = Z | S Nat
+
+  data Vec :: Nat -> Type -> Type where
+    VNil :: Vec Z a
+    (:>) :: { head :: a, tail :: Vec n a } -> Vec (S n) a
+  |])

